ckks/encoder.go

```diff
@@ -132,88 +125,129 @@ func NewEncoder(params Parameters) Encoder {
-	return encoder.EncodeAtLvlNew(encoder.params.MaxLevel(), values, logSlots)
-}
-
-// EncodeAtLvlNew encodes a slice of complex128 of length slots = 2^{logSlots} on new plaintext at the desired level.
-func (encoder *encoderComplex128) EncodeAtLvlNew(level int, values []complex128, logSlots int) (plaintext *Plaintext) {
-	plaintext = NewPlaintext(encoder.params, level, encoder.params.Scale())
-	encoder.Encode(plaintext, values, logSlots)
-	return
-}
-
-// Encode encodes a slice of complex128 of length slots = 2^{logSlots} on the input plaintext.
-func (encoder *encoderComplex128) Encode(plaintext *Plaintext, values []complex128, logSlots int) {
-	encoder.Embed(values, logSlots)
-	encoder.ScaleUp(plaintext.Value, plaintext.Scale, encoder.params.RingQ().Modulus[:plaintext.Level()+1])
-	plaintext.Value.IsNTT = false
-}
-
 // EncodeNTTNew encodes a slice of complex128 of length slots = 2^{logSlots} on new plaintext at the maximum level.
 // Returns a plaintext in the NTT domain.
-func (encoder *encoderComplex128) EncodeNTTNew(values []complex128, logSlots int) (plaintext *Plaintext) {
-	return encoder.EncodeNTTAtLvlNew(encoder.params.MaxLevel(), values, logSlots)
+func (encoder *encoderComplex128) EncodeNew(values interface{}, logSlots int) (plaintext *Plaintext) {
+	return encoder.EncodeAtLvlNew(encoder.params.MaxLevel(), values, logSlots)
 }
 
 // EncodeNTTAtLvlNew encodes a slice of complex128 of length slots = 2^{logSlots} on new plaintext at the desired level.
 // Returns a plaintext in the NTT domain.
-func (encoder *encoderComplex128) EncodeNTTAtLvlNew(level int, values []complex128, logSlots int) (plaintext *Plaintext) {
+func (encoder *encoderComplex128) EncodeAtLvlNew(level int, values interface{}, logSlots int) (plaintext *Plaintext) {
 	plaintext = NewPlaintext(encoder.params, level, encoder.params.Scale())
-	encoder.EncodeNTT(plaintext, values, logSlots)
+	encoder.Encode(plaintext, values, logSlots)
 	return
 }
 
 // EncodeNTT encodes a slice of complex128 of length slots = 2^{logSlots} on the input plaintext.
 // Returns a plaintext in the NTT domain.
-func (encoder *encoderComplex128) EncodeNTT(plaintext *Plaintext, values []complex128, logSlots int) {
-	encoder.Encode(plaintext, values, logSlots)
+func (encoder *encoderComplex128) Encode(plaintext *Plaintext, values interface{}, logSlots int) {
+	encoder.embed(values, logSlots)
+	scaleUpVecExact(encoder.valuesFloat[:encoder.params.N()], plaintext.Scale, encoder.params.RingQ().Modulus[:plaintext.Level()+1], plaintext.Value.Coeffs)
 	encoder.params.RingQ().NTTLvl(plaintext.Level(), plaintext.Value, plaintext.Value)
 	plaintext.Value.IsNTT = true
 }
 
 // Embed encodes a vector and stores internally the encoded values.
 // To be used in conjunction with ScaleUp.
-func (encoder *encoderComplex128) Embed(values []complex128, logSlots int) {
+func (encoder *encoderComplex128) embed(values interface{}, logSlots int) {
 
 	encoder.WipeInternalMemory()
```

Comment:
Agree

Reply:
The embed method knows the size of the input so it can write 0s where needed. This avoid writing 0s just to overwrite with the inputs later on. 

In general, the `WipeInternalMemory` method is problematic: the user should not have to care about the internal state of lattigo structs.